poc/vdaf_mastic.py

```diff
@@ -0,0 +1,370 @@
+import sys
+sys.path.append('draft-irtf-cfrg-vdaf/poc')
+
+from common import Unsigned, front, vec_add, vec_sub
+from flp_generic import FlpGeneric
+import hashlib
+from typing import Optional, Union
+from vdaf import Vdaf, test_vdaf
+from vdaf_poplar1 import Poplar1
+from vidpf import Vidpf
+from xof import XofShake128
+
+
+USAGE_QUERY_RAND = 0
+USAGE_PROVE_RAND = 1
+USAGE_PROOF_SHARE = 2
+
+
+class Mastic(Vdaf):
+    # Operational types and parameters.
+    Field = None # Set by `with_params()`
+    Vidpf = None # Set by `with_params()`
+    Flp = None # Set by `with_params()`
+    Xof = XofShake128
+    ROOT_PROOF = hashlib.sha3_256().digest() # Hash of the empty string
+
+    # Parameters required by `Vdaf`.
+    ID: Unsigned = 0xFFFFFFFF
+    VERIFY_KEY_SIZE = Xof.SEED_SIZE
+    NONCE_SIZE = 16
+    RAND_SIZE = None # Set by `with_params()`
+    SHARES = 2
+    ROUNDS = 1
+
+    # Types required by `Vdaf`
+    AggParam = Poplar1.AggParam
+    PublicShare = None # TODO(cjpatton)
+    InputShare = None # TODO(cjpatton)
+    PrepMessage = None
+
+    @classmethod
+    def shard(cls, measurement, nonce, rand):
+        (vidpf_rand, rand) = front(cls.Vidpf.RAND_SIZE, rand)
+        (prove_rand_seed, rand) = front(cls.Xof.SEED_SIZE, rand)
+        (helper_seed, rand) = front(cls.Xof.SEED_SIZE, rand)
+
+        (alpha, meas) = measurement
+        beta = cls.Flp.encode(meas)
+
+        # Generate VIDPF keys.
+        (init_seed, correction_words, cs_proofs) = \
+            cls.Vidpf.gen(alpha, beta, nonce, vidpf_rand)
+        public_share = (correction_words, cs_proofs)
+
+        # Generate FLP shares.
+        prove_rand = cls.Xof.expand_into_vec(cls.Field,
+            prove_rand_seed,
+            cls.domain_separation_tag(USAGE_PROVE_RAND),
+            b'',
+            cls.Flp.PROVE_RAND_LEN,
+        )
+
+        proof = cls.Flp.prove(beta, prove_rand, [])
+        leader_proof_share = vec_sub(proof,
+                                     cls.helper_proof_share(helper_seed))
+
+        input_shares = [
+            (init_seed[0], leader_proof_share),
+            (init_seed[1], helper_seed),
+        ]
+        return (public_share, input_shares)
+
+    @classmethod
```

Comment:
Who runs `is_valid()`? I'm a bit confused about what it checks.

Reply:
See https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-vdaf-07#section-5.3